### PR TITLE
Use population standard deviation as confidence value

### LINKED ISSUE #4: Check that confidence calculation matches our latest discussions
@audeerington we had a long discussion for one of our databases what should be the best method to calculate confidence values for regression values. 

Could you check that the implementation provided in this package at https://github.com/audeering/audpsychometric/blob/master/audpsychometric/core/gold_standard.py#L30-L33
matches on what we had agreed on there.

/cc @sandneurodata

### PR DESCRIPTION
**Pull request by sandneurodata**
Closes #4.

Updates confidence calculation to use the population standard deviation, compare [our internal discussion](https://gitlab.audeering.com/data/msppodcast/-/merge_requests/23#note_196384).

The code for the confidence values can be given by:

```python
def confidence(rating, minimum, maximum):
    max_std = (maximum - minimum) / 2  # == np.std([minimum, maximum], ddof=0)
    std = np.std(rating, ddof=0)
    std_norm = np.clip(std/max_std, 0, 1)
    return 1 - std_norm
```

In addition, the following changes are added:
* Add `minimum` and `maximum` to all functions that calculate confidence as those values are required now
* Add `axis` with default value `1` to all functions calculating an average to allow to switch to having rater as index (axis=0)
* Correct calculation of confidence for numerical values (mean and median) by removing the added `'gold'` column from the dataframe (before the confidence was calculated over all ratings + gold standard)
* Replaced the usage of `df_holzinger_swineford` data with user defined examples in the tests for confidences

---

This does not add documentation on confidence calculation to the docstrings as we want to separate the confidence calculation into separate functions anyway in another pull request (see #7). I would then add the documentation to that pull request.

### CODE REVIEW

**Comment by hagenw** [comment]
As you can change the code without triggering a test to fail it looks like a test for the confidence value needs to be added/adjusted as well.

**Comment by sandneurodata** [comment]
> As you can change the code without triggering a test to fail it looks like a test for the confidence value needs to be added/adjusted as well.

Isn't the confidence value obtained from _confidence_numerical already tested in the following test:s def test_mean_based_gold_standard(df_holzinger_swineford) and def test_median_based_gold_standard(df_holzinger_swineford) from test_gold_standard.py?

**Comment by hagenw** [comment]
> test_mean_based_gold_standard

It might be tested, but your changes to the code should change the resulting confidence values. And I would expect that there exists a test that checks that we get the correct confidence values. E.g. for a dummy scenario for which we calculated the confidence values.

**Comment by sandneurodata** [comment]
ok, i got it - do you know where i could get such dummy data? i could create some and calculate the confidence value with the function included here but that wouldn't serve a purpose since one wants to make sure that the computation we have here is the correct one, right?

**Comment by hagenw** [comment]
I would try to create the data manually and make some simple tests, e.g. in https://github.com/audeering/audpsychometric/issues/4 it is stated that 
> the std for a sample with ratings [0,1] would be the same as for another sample with ratings [0,0,1,1]
This could be one test.
I guess the max cut off could then be tested by changing `1` to a different value.

**Comment by sandneurodata** [comment]
I created a test for the numerical confidence and it is actually failing and i don't know why... when I ran the operation myself it is fine, but when it's running in a function, it is not anymore and I quite can't get it... Here is the code:

def test_numerical_confidence():
    """test that numerical confidence computation is appropriate"""

    row_1 = pd.Series([0, 1], copy=False)
    row_2 = pd.Series([0, 0, 1, 1], copy=False)
    assert _confidence_numerical(row_1) == _confidence_numerical(row_2)

From the discussion in #4 i would suspect those to be equal but no...
i would appreciate if you could tell me what goes wrong here - thanks a lot!

**Comment by hagenw** [comment]
They return indeed different results:

```python
>>> _confidence_numerical(row_1)
0.717157287525381
>>> _confidence_numerical(row_2)
0.7690598923241497
```
so it's not a numerical problem.

It seems that the current implementation is not corect.
I used the following code for a dataframe, and it indeed produces the same result:

```python
def confidence(df):
    std = df.std(axis=1, ddof=0)
    cutoff_max = df.max(axis=1) / df.mean(axis=1)
    std_norm = 1 - std / cutoff_max
    return std_norm.clip(lower=0)
```
and then
```python
>>> confidence(pd.DataFrame([row_1]))
0    0.75
dtype: float64
>>> confidence(pd.DataFrame([row_2]))
0    0.75
dtype: float64
```

but the result of 0.75 looks a little bit strange to me. For ratings of `[0, 1]` I would have expected that the confidence should be 0.0 instead.

@audeerington anything we miss here?

**Comment by hagenw** [comment]
I found my error, it should be:

```python
def confidence(df):
    std = df.std(axis=1, ddof=0)
    cutoff_max = df.max(axis=1) - df.mean(axis=1)
    std_norm = 1 - std / cutoff_max
    return std_norm.clip(lower=0)
```
then it will indeed return 0.0 in both cases.

**Comment by audeerington** [comment]
One other point that comes to mind is that it is possible that a dataframe doesn't contain the true maximum and minimum that a database permits. So I guess that either should to be explained to the user, or there should be some argument where you can manually set the range of permitted ratings.

**Comment by audeerington** [comment]
Actually, I realized that if you calculate the maximum and mean for each rating, then the result will not be correct.

The formula I wrote before should give you a single value for the maximum std that can occur for ratings within a set range. So by calculating the maximum and mean for each row separately, you will get totally different results than if you normalize with one and the same value.

**Comment by hagenw** [comment]
> To get the maximum population standard deviation for ratings in $[x_0,x_1]$, we can use the formula $x_1-\bar{x}$. This is the value we decided to use for the cutoff.

This means the mean in that formula refers just to the mean of the range values, correct?

$x_1-\bar{x} = x_1 - \frac{1}{2} (x_0 + x_1)$

Which means if we have rating in the range $[0, 1]$, the cutoff value would be $0.5$, right?

**Comment by audeerington** [comment]
Yes, that is exactly correct. And $0.5$ is the maximum value for the population std in the range $[0,1]$, which is why we use that value to normalize the std of all samples.

**Comment by hagenw** [comment]
An implementation that works on a Series would then look like this:

```python
def confidence(y, minimum, maximum):
    cutoff = maximum - 1 / 2 * (minimum + maximum)
    std = y.std(ddof=0)
    return max([0, 1 - std / cutoff])
```

This means we need to provide `_confidence_numerical()` with `minimum` and `maximum` arguments.
Instead of adding this to all the gold standard function it might be better to also split up the calculation of gold standard and confidence as proposed in https://github.com/audeering/audpsychometric/issues/7

**Comment by sandneurodata** [comment]
sorry, i have a naive question - if i understood correctly, the function you proposed to calculate the confidence would require for each serie a different minimum and maximum values, right? this means that if one wanted to call such function from let's say def gold_standard_mean(), for each row of the dataframe one would have to give a different minimum and maximum? or a fixed minimum and maximum would work for the whole dataframe?

**Comment by hagenw** [comment]
Minimum and maximum should be fixed for all ratings for a given question/scheme.
So in your case it would be fixed for the whole dataframe.

I just provided my code for a Series as you used a Series as input for `_confidence_numerical()`.

**Comment by sandneurodata** [comment]
> Minimum and maximum should be fixed for all ratings for a given question/scheme. So in your case it would be fixed for the whole dataframe.
> 
> I just provided my code for a Series as you used a Series as input for `_confidence_numerical()`.

ok, thanks. One could eventually just provide the min and max as arguments in gold_standard_mean for instance, instead of splitting the gold standard and confidence into 2 separated functions?

**Comment by hagenw** [comment]
> > Minimum and maximum should be fixed for all ratings for a given question/scheme. So in your case it would be fixed for the whole dataframe.
> > I just provided my code for a Series as you used a Series as input for `_confidence_numerical()`.
> 
> ok, thanks. One could eventually just provide the min and max as arguments in gold_standard_mean for instance, instead of splitting the gold standard and confidence into 2 separated functions?

Yes, to not over complicate things, I would indeed not mix those two things here.
I would indeed start adding the additional `min`, `max` arguments (or maybe `minimum`, `maximum` to use the same wording as in `audformat.Scheme`) in this pull request.

If we then split up the confidence calculation from the gold standard calculation we can decide/implement afterwards.

**Comment by sandneurodata** [comment]
I've just implemented the minimum/maximum argument everywhere it needed it but here the tests are failing (i.e the coverage) whereas everything is fine when i run them locally... i must have missed something somewhere but what??

**Comment by hagenw** [comment]
Did you maybe forgot to push changes to the tests, because the tests are failing as the new `minimum`, `maximum` arguments are not provided, e.g.

image

**Comment by sandneurodata** [comment]
yes, indeed, i forgot to change it there :-)

**Comment by sandneurodata** [comment]
should i put a fixed value there or just define the arguments? thanks

**Comment by hagenw** [comment]
One way to define test in `pytest` with arguments is to use `@pytest.mark.parametrize()`, this way you can define a list of input args and expected outputs. You could start with a single example for `minimum` and `maximum`, but you could also add different values.

If you are able to define an expected result, it could be done similar to
```python
@pytest.mark.parametrize(
    'df, minimum, maximum, expected_df',
    [
        (
            df_holzinger_swineford,
            0,
            1,
            expected_df,
        ),
    ]
)
def test_mean_based_gold_standard(df, minimum, maximum, expected_df):
    """Happy Flow test for mode based gold standard"""
    df = audpsychometric.gold_standard_mean(df, minimum, maximum)
    pd.testing.assert_frame_equal(df, expected_df)
```

Otherwise you need to adjust the tests by e.g. just providing `df, minimum, maximum` as inputs and then check for the stuff you want inside the test.

**Comment by sandneurodata** [comment]
> One way to define test in `pytest` with arguments is to use `@pytest.mark.parametrize()`, this way you can define a list of input args and expected outputs. You could start with a single example for `minimum` and `maximum`, but you could also add different values.
> 
> If you are able to define an expected result, it could be done similar to
> 
> ```python
> @pytest.mark.parametrize(
>     'df, minimum, maximum, expected_df',
>     [
>         (
>             df_holzinger_swineford,
>             0,
>             1,
>             expected_df,
>         ),
>     ]
> )
> def test_mean_based_gold_standard(df, minimum, maximum, expected_df):
>     """Happy Flow test for mode based gold standard"""
>     df = audpsychometric.gold_standard_mean(df, minimum, maximum)
>     pd.testing.assert_frame_equal(df, expected_df)
> ```
> 
> Otherwise you need to adjust the tests by e.g. just providing `df, minimum, maximum` as inputs and then check for the stuff you want inside the test.

sorry, was thinking about something... what about calculating instead the minimum and maximum of the whole dataframe instead of defining ourselves the values? Since the whole dataframe is an argument for the gold standard calculation, it is quite straitforward - unless there is a advantage of defining it ourselves but then I am not sure i get it...

**Comment by hagenw** [comment]
> what about calculating instead the minimum and maximum of the whole dataframe instead

This would be another approach. We might think about if we want to add this as an option, but I'm not sure we need/want this yet.

E.g. let's assume you annotate on a scale from 0 to 10. It might be that your minimum response is 2 in the whole database, so you will get  a different result then.

**Comment by sandneurodata** [comment]
ok, i got it - thanks! will fix this then appropriately

**Comment by sandneurodata** [comment]
I wrote a fixture in the conftest.py file that creates the expected gold standard and confidence from holzinger_swineford and then i implemented what you are suggested but it doesn't work (apparently it cannot access the expected_df and also now the df_holzinger_swineford) - what did i do wrong? thanks again for your help

**Comment by hagenw** [comment]
@ChristianGeng I updated the code, the tests, and the description of this pull request. It should now be ready for review.

File: tests/test_gold_standard.py [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+        (
+            pd.DataFrame(
+                [
+                    [1, 2, 3],
+                    [3, 4, 5],
+                ]
+            ),
+            0,
+            10,
+            0,
```
[the remark this answers is not in the document]
Comment (hagenw) [reply tests/test_gold_standard.py:169]: The main motivation is to support different orientations of the dataframe:

```python
>>> df1 = pd.DataFrame([[0, 1], [0.5, 1.5]], columns=['rater1', 'rater2'])
>>> df1
   rater1  rater2
0     0.0     1.0
1     0.5     1.5
>>> df2 = pd.DataFrame([[0, 0.5], [1.0, 1.5]], index=pd.Index(['rater1', 'rater2']))
>>> df2
          0    1
rater1  0.0  0.5
rater2  1.0  1.5
```

Now we can simply do:

```python
>>> audpsychometric.gold_standard_mean(df1, 0, 2)
   gold_standard  confidence
0            0.5         0.5
1            1.0         0.5
>>> audpsychometric.gold_standard_mean(df2, 0, 2, axis=0)
   gold_standard  confidence
0            0.5         0.5
1            1.0         0.5
```

File: audpsychometric/core/gold_standard.py [hunk 2]
```diff
@@ -147,29 +192,38 @@ def _confidence_categorical(row: pd.Series) -> float:
...
+        maximum: maximum value of the data to calculate cut off
 
     Returns:
         numerical confidence score
 
     """
     raters = row.index.tolist()
-    return max([0., 1 - row[raters].std() / cutoff_max])
+    cutoff_max = maximum - 1 / 2 * (minimum + maximum)
+    return max([0., 1 - row[raters].std(ddof=0) / cutoff_max])
```
[the remark this answers is not in the document]
Comment (hagenw) [reply audpsychometric/core/gold_standard.py:219]: No, `ddof=0` is [not the default](https://pandas.pydata.org/docs/reference/api/pandas.Series.std.html):

image

**Pull request merged**